### PR TITLE
parsing: Add `key` to matrix stages

### LINKED ISSUE #10043: Add `key` to matrix
The addition of `matrix` functionality simplified quite a lot for me. However, contrary to `foreach` stages, there is no `key` property available. I often find myself composing output paths using the parametrization fields.

Let me give an example – first, using foreach:

```yaml
train_model:
  foreach:
    backbone_A_dim_100:
      backbone: A
      dim: 100
   backbone_B_dim_100:
     backbone: B
     dim: 100
   backbone_A_dim_200:
     ....
  do:
    cmd: python train_model.py --backbone ${item.backbone} --dim ${item.dim}
    outs:
    - data/models/${key}/
```

This can be simplified into a `matrix` stage:

```yaml
train_model:
  matrix:
    backbone: ["A", "B"]
   dim: [100, 200]
  cmd: python train_model.py --backbone ${item.backbone} --dim ${item.dim}
  outs:
  - data/models/${item.backbone}_${item.dim}/
```

The output path is composed of both the `item.backbone` and `item.dim`, which in this simplified case might not be that problematic, but it becomes more demanding and error prone with more `matrix` parameters. An easier way would be to include the `key` of the `matrix` stage into the templating context, i.e.:

```yaml
train_model:
  matrix:
    backbone: ["A", "B"]
   dim: [100, 200]
  cmd: python train_model.py --backbone ${item.backbone} --dim ${item.dim}
  outs:
  - data/models/${key}/
```

I already implemented the requested functionality ~~and will link it in a moment.~~ – see [PR#10044](https://github.com/iterative/dvc/pull/10044)

### PR DESCRIPTION
**Pull request by pbielak**
This commit adds the `key` property to the templaing context of `matrix` stages. Similarly to `foreach` stages, the `key` is available using `${key}` substitution.

Fixes #10043

* [x] ❗ I have followed the [Contributing to DVC](https://dvc.org/doc/user-guide/contributing/core) checklist.

* [x] 📖 If this PR requires [documentation](https://dvc.org/doc) updates, I have created a separate PR (or issue, at least) in [dvc.org](https://github.com/iterative/dvc.org) and linked it here.

Thank you for the contribution - we'll try to review it as soon as possible. 🙏

### CODE REVIEW

File: dvc/parsing/__init__.py [hunk 1]
```diff
@@ -499,8 +499,26 @@ def _resolve_matrix_data(self) -> Dict[str, List]:
             iterable = self.context.resolve(self.matrix_data, unwrap=False)
         except (ContextError, ParseError) as exc:
             format_and_raise(exc, f"'{self.where}.{self.name}.matrix'", self.relpath)
+
+        # Matrix entries will have `key` and `item` added to the context.
+        # Warn users if these are already in the context from the global vars.
+        self._warn_if_overwriting(["key", "item"])
```
Comment (skshetry) [inline dvc/parsing/__init__.py:505]: Let's not hardcode these here.

```suggestion
        self._warn_if_overwriting([self.pair, self.value])
```

File: tests/func/parsing/test_matrix.py [hunk 2]
```diff
@@ -50,3 +50,47 @@ def test_matrix_interpolated(tmp_dir, dvc, matrix):
...
+            "win-3.8-dict1-list0",
+            "win-3.8-dict1-list1",
+            "linux-3.7-dict0-list0",
+            "linux-3.7-dict0-list1",
+            "linux-3.7-dict1-list0",
+            "linux-3.7-dict1-list1",
+            "linux-3.8-dict0-list0",
+            "linux-3.8-dict0-list1",
+            "linux-3.8-dict1-list0",
+            "linux-3.8-dict1-list1",
```
Comment (skshetry) [inline tests/func/parsing/test_matrix.py:94]: Do you mind unrolling this loop, please? That might make it easier to see when things fail.

Comment (pbielak) [reply tests/func/parsing/test_matrix.py]: Done

**Comment by skshetry** [comment]
I see that there is a request for supporting dictionary inside matrix instead of a list in discord. Foreach provides key of the dictionary as a `key`, and value as `item`.

What would be the way to access those keys in `matrix` if we ever support that? So I think we need to clarify that first. cc @dberenbaum.

One way might be to not support dict at all, and redirect users to use a list instead.

**Comment by dberenbaum** [comment]
@skshetry Was this a documented feature of `foreach`? I can't seem to find it in the docs and don't remember seeing it before.

**Comment by pbielak** [comment]
@dberenbaum Yes, it is documented - https://dvc.org/doc/user-guide/project-structure/dvcyaml-files#foreach-stages

```
DVC can also iterate on a dictionary given directly to foreach, resulting in two substitution expressions being available: ${key} and ${item}. The former is used for the stage names:
```

**Comment by dberenbaum** [comment]
Sorry, don't know how I missed that. 

@skshetry I'm also not sure if we should support dict for matrix. I favor being more flexible, but if it doesn't enable any additional scenarios and increases complexity/confusion, it might not be worth it.

If we do, I guess it should be possible to support this PR and have dict support like this:

```yaml
train_model:
  matrix:
    backbone:
      A:
       arch: resnet
      B:
        arch: alexnet
    dim: [100, 200]
  cmd: python train_model.py --backbone ${key.backbone} --arch ${item.backbone.arch} --dim ${item.dim}
  outs:
  - data/models/${key}/
```

WDYT?

**Review by dberenbaum — APPROVED** [review]
LGTM, thanks @pbielak! Would you also mind doing a docs PR to explain the new behavior?

**Pull request merged**